### PR TITLE
Allow adding tags to existing conversations

### PR DESCRIPTION
**Pull request by swarelle**
Added a text box and POST form on chat.jsp to allow for adding tags to an existing conversation. Also added a line under the chat title to display current tags. Will update in a separate branch to make those tags links to a page with all conversations with that tag.

### CODE REVIEW

File: src/main/java/codeu/controller/ChatServlet.java [hunk 1]
```diff
@@ -103,6 +105,26 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
 
     List<Message> messages = messageStore.getMessagesInConversation(conversationId);
 
+    String username = (String) request.getSession().getAttribute("user");
```
Comment (joycezhao2) [inline src/main/java/codeu/controller/ChatServlet.java]: can you add comments along the code to document what's happening?

File: src/main/java/codeu/controller/ChatServlet.java [hunk 2]
```diff
@@ -103,6 +105,26 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+    if (username != null) {
+      User user = userStore.getUser(username);
+      if (user != null) {
+        List<Tag> tags = user.getConversationTags(conversation);
```
Comment (swarelle) [inline src/main/java/codeu/controller/ChatServlet.java]: Directly get list of Tags from Conversation object

File: src/main/java/codeu/controller/TagServlet.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    // Get all tags
+    String tags = request.getParameter("tags");
+    if (tags != null && tags.length() > 0) {
+      if (!tags.matches("([\\w*](, ))*[\\w*]*")) {
+        request.setAttribute("error", "Please enter tags as comma-separated words with one space between them. Tags can only contain letters and numbers.");
+        request.setAttribute("conversation", conversation);
+        request.setAttribute("messages", messageStore.getMessagesInConversation(conversation.getId()));
+        List<Tag> tagsList = user.getConversationTags(conversation);
```
Comment (swarelle) [inline src/main/java/codeu/controller/TagServlet.java]: Update to be not user-specific

File: src/main/java/codeu/controller/TagServlet.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        }
+        request.getRequestDispatcher("/WEB-INF/view/chat.jsp").forward(request, response);
+        return;
+      }
+
+      // Add all tags to the user's tagged conversations Map
+      List<String> splitTags = Arrays.asList(tags.split(", "));
+      for (String tag : splitTags) {
+        Tag newTag = new Tag(UUID.randomUUID(), tag, Instant.now());
+        user.addTaggedConversation(newTag, conversation);
```
Comment (swarelle) [inline src/main/java/codeu/controller/TagServlet.java]: Change to be non user-specific

File: src/main/java/codeu/controller/TagServlet.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      }
+      userStore.putUser(user);
```
Comment (swarelle) [inline src/main/java/codeu/controller/TagServlet.java]: Update to put Tag and Conversation in their Stores

File: src/main/java/codeu/controller/TagServlet.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+          request.setAttribute("tags", "");
+        }
+        request.getRequestDispatcher("/WEB-INF/view/chat.jsp").forward(request, response);
+        return;
+      }
+
+      // Add all tags to the user's tagged conversations Map
+      List<String> splitTags = Arrays.asList(tags.split(", "));
+      for (String tag : splitTags) {
+        Tag newTag = new Tag(UUID.randomUUID(), tag, Instant.now());
```
Comment (joycezhao2) [inline src/main/java/codeu/controller/TagServlet.java]: check if tag exists (can be in a new PR after TagStore is merged)

File: src/main/java/codeu/model/data/User.java [hunk 7]
```diff
@@ -56,8 +66,53 @@ public User(UUID id, String name, Instant creation, String hashedPassword) {
...
+      tags.add(tag);
+      conversationTags.put(conversation, tags);
+    }
+  }
+
+  /**
+   * Gets an alphabetically sorted list of tags this User has added for a 
+   * particular conversation.
+   */
+  public List<Tag> getConversationTags(Conversation conversation) {
```
Comment (swarelle) [inline src/main/java/codeu/model/data/User.java]: Move method to Conversation object -- now Conversations hold the List of Tags

File: src/main/java/codeu/controller/ChatServlet.java [hunk 8]
```diff
@@ -103,6 +105,26 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
...
+        if (tags.size() > 0) {
+          StringBuilder sb = new StringBuilder();
+          for (Tag tag : tags) {
+            sb.append(tag.getTag());
+            sb.append(", ");
+          }
+          request.setAttribute("tags", sb.toString().substring(0, sb.length() - 2));
```
Comment (joycezhao2) [inline src/main/java/codeu/controller/ChatServlet.java]: instead set attribute as list of tags

File: src/main/webapp/WEB-INF/view/chat.jsp [hunk 9]
```diff
@@ -56,10 +57,29 @@ User loggedInUser = UserStore.getInstance().getUser(user);
...
+        <h2 style="color:red"><%= request.getAttribute("error") %></h2>
+    <% } %>
+
     <h1><%= conversation.getTitle() %>
       <a href="" style="float: right">&#8635;</a></h1>
-
+    
+    <% if (loggedInUser != null) { %>
+        <label class="form-control-label">Current tags:</label>
+        <p><%= request.getAttribute("tags") %><p>
```
Comment (swarelle) [inline src/main/webapp/WEB-INF/view/chat.jsp]: Iterate through List of Tags to make each of them a link to their "Tag profile page"

File: src/main/java/codeu/model/data/Tag.java [hunk 10]
```diff
@@ -1,25 +1,41 @@
...
    */
-  public Tag(UUID id, String tag, Instant creation) {
+  public Tag(UUID id, Conversation conversation, String tag, Instant creation) {
     this.id = id;
     this.tag = tag;
     this.creation = creation;
+    this.conversations = new ArrayList<>();
+    this.conversations.add(conversation);
+  }
+
```
Comment (masterford) [inline src/main/java/codeu/model/data/Tag.java]: Is there an advantage with taking in a conversation object for a new tag as opposed to a conversation ID?

**Pull request merged**